File: binary-search/binary_search-duplicadas.py
def bin_search(arr, tamanho, target):
    inicio = 0
    fim = tamanho - 1
    resultado = -1

    while(inicio <= fim):
        meio = inicio + (fim - inicio) // 2

        if arr[meio] == target:
            resultado = meio
            fim = meio - 1
        elif arr[meio] > target:
            fim = meio - 1
        else:
            inicio = meio + 1
    return resultado

arr = [1,2,2,2,3,3,4]
target = int(input("Qual elemento deseja saber o indice da primeira ocorrencia: "))

resultado = bin_search(arr, len(arr), target)

if resultado == -1: 
    print("Elemento nao encontrado!")
else:
    print("Elemento encontrado na posicao", resultado)

File: binary-search/binary_search-duplicadas.c
#include <stdio.h>

int bin_search(int *arr, int n, int x){
    int inicio = 0;
    int fim = n - 1;
    int meio;
    int resultado = -1;

    while(inicio <= fim){
        meio = inicio + (fim - inicio) / 2;

        if(arr[meio] == x){
            resultado = meio;
            fim = meio - 1;
        }else if(arr[meio] > x){
            fim = meio - 1;
        } else {
            inicio = meio + 1;
        }
    }
    return resultado;
}

int main(){

    int arr[] = {1,2,2,2,3,3,4};
    int tamanho = sizeof(arr) / sizeof(arr[0]);

    int elemento;
    printf("Qual elemento deseja procurar: ");
    scanf("%d", &elemento);

    int resultado = bin_search(arr, tamanho, elemento);

    if(resultado == -1){
        printf("Elemento nao encontrado!\n");
    } else {
        printf("A primeira ocorrencia eh no indice %d\n", resultado);
    }

    return 0;
}
